File: cowork-roadmap/src/main/java/com/cowork/roadmap/global/team/ProjectionTopicGenerationRegistry.java
package com.cowork.roadmap.global.team;

import java.util.concurrent.atomic.AtomicReference;

import org.springframework.stereotype.Component;

@Component
public class ProjectionTopicGenerationRegistry {

    private final AtomicReference<String> topicId = new AtomicReference<>();

    void markAssigned(String currentTopicId) {
        topicId.set(currentTopicId);
    }

    void markRevoked() {
        topicId.set(null);
    }

    String requireTopicId() {
        String current = topicId.get();
        if (current == null) {
            throw new IllegalStateException("Kafka projection topic generation is not initialized");
        }
        return current;
    }
}

File: cowork-channel/src/main/kotlin/com/cowork/channel/global/projection/ProjectionTopicGenerationRegistry.kt
package com.cowork.channel.global.projection

import org.springframework.stereotype.Component
import java.util.concurrent.ConcurrentHashMap

@Component
class ProjectionTopicGenerationRegistry {
    private val topicIds = ConcurrentHashMap<ProjectionStream, String>()

    fun markAssigned(stream: ProjectionStream, topicId: String) {
        topicIds[stream] = topicId
    }

    fun markRevoked(stream: ProjectionStream) {
        topicIds.remove(stream)
    }

    fun topicId(stream: ProjectionStream): String? = topicIds[stream]
}
